### PR TITLE
Compile without libssh

### PR DESCRIPTION
**Pull request by Yaraslaut**

### CODE REVIEW

File: src/vtpty/CMakeLists.txt [hunk 1]
```diff
@@ -42,8 +44,7 @@ else()
     endif()
 endif()
 if(_include_SshSession_module)
-    list(APPEND vtpty_SOURCES SshSession.cpp)
-    list(APPEND vtpty_HEADERS SshSession.h)
+    add_compile_definitions("SSH")
```
Comment (christianparpart) [inline src/vtpty/CMakeLists.txt:47]: Hmm... if we don't have libssh2 library/headers, we should not build with SshSession at all. what was the problem with libssh2 not being present? What failed?

Comment (Yaraslaut) [reply src/vtpty/CMakeLists.txt:47]: > Hmm... if we don't have libssh2 library/headers, we should not build with SshSession at all. what was the problem with libssh2 not being present? What failed?

Config loading

Comment (Yaraslaut) [reply src/vtpty/CMakeLists.txt:47]: > > Hmm... if we don't have libssh2 library/headers, we should not build with SshSession at all. what was the problem with libssh2 not being present? What failed?
> 
> Config loading

Initially, but we use SshSession object in some other places like sessions initializatoin and others

Comment (christianparpart) [reply src/vtpty/CMakeLists.txt:47]: What do you think about the following proposal instead?

```diff
diff --git a/src/contour/Config.cpp b/src/contour/Config.cpp
index 14deab80..d76ed8e5 100644
--- a/src/contour/Config.cpp
+++ b/src/contour/Config.cpp
@@ -1403,6 +1403,7 @@ namespace
             loginShell.erase(loginShell.begin());
             terminalProfile.shell.arguments = loginShell;
         }
+        #if defined(VTPTY_LIBSSH2)
         if (auto const sshNode = profile["ssh"]; sshNode && sshNode.IsMap())
         {
             usedKeys.emplace(fmt::format("{}.{}.ssh", parentPath, profileName));
@@ -1442,6 +1443,7 @@ namespace
             else if (auto const p = Process::homeDirectory() / ".ssh" / "known_hosts"; fs::exists(p))
                 ssh.knownHostsFile = p;
         }
+        #endif

         tryLoadChildRelative(usedKeys, profile, basePath, "maximized", terminalProfile.maximized, logger);
         tryLoadChildRelative(usedKeys, profile, basePath, "fullscreen", terminalProfile.fullscreen, logger);
@@ -1537,7 +1539,9 @@ namespace
             terminalProfile.shell.env["COLORTERM"] = "truecolor";

         // This is currently duplicated. Environment vars belong to each parent struct, not just shell.
+        #if defined(VTPTY_LIBSSH2)
         terminalProfile.ssh.env = terminalProfile.shell.env;
+        #endif
         // }}}

         strValue = fmt::format("{}", terminalProfile.terminalId);
diff --git a/src/contour/ContourGuiApp.cpp b/src/contour/ContourGuiApp.cpp
index 0a17d2dd..3c9b659f 100644
--- a/src/contour/ContourGuiApp.cpp
+++ b/src/contour/ContourGuiApp.cpp
@@ -38,7 +38,10 @@ using std::string_view;
 using std::vector;

 using vtpty::Process;
+
+#if defined(VTPTY_LIBSSH2)
 using vtpty::SshSession;
+#endif

 using namespace std::string_literals;

@@ -178,8 +181,10 @@ void ContourGuiApp::onExit(TerminalSession& session)
 {
     if (auto const* localProcess = dynamic_cast<vtpty::Process const*>(&session.terminal().device()))
         _exitStatus = localProcess->checkStatus();
+#if defined(VTPTY_LIBSSH2)
     else if (auto const* sshSession = dynamic_cast<vtpty::SshSession const*>(&session.terminal().device()))
         _exitStatus = sshSession->exitStatus();
+#endif
 }

 QUrl ContourGuiApp::resolveResource(std::string_view path)
@@ -418,6 +423,7 @@ int ContourGuiApp::terminalGuiAction()

     if (_exitStatus.has_value())
     {
+#if defined(VTPTY_LIBSSH2)
         if (holds_alternative<SshSession::ExitStatus>(*_exitStatus))
         {
             auto const sshExitStatus = get<SshSession::ExitStatus>(*_exitStatus);
@@ -426,7 +432,9 @@ int ContourGuiApp::terminalGuiAction()
             else if (holds_alternative<SshSession::SignalExit>(sshExitStatus))
                 rv = EXIT_FAILURE;
         }
-        else if (holds_alternative<Process::ExitStatus>(*_exitStatus))
+        else
+#endif
+            if (holds_alternative<Process::ExitStatus>(*_exitStatus))
         {
             auto const processExitStatus = get<Process::ExitStatus>(*_exitStatus);
             if (holds_alternative<Process::NormalExit>(processExitStatus))
diff --git a/src/contour/TerminalSession.cpp b/src/contour/TerminalSession.cpp
index 241737c3..fd1f6426 100644
--- a/src/contour/TerminalSession.cpp
+++ b/src/contour/TerminalSession.cpp
@@ -585,6 +585,7 @@ void TerminalSession::onClosed()
         else
             sessionLog()("Process terminated after {} seconds.", diff.count());
     }
+#if defined(VTPTY_LIBSSH2)
     else if (auto* sshSession = dynamic_cast<vtpty::SshSession*>(&_terminal.device()))
     {
         auto const exitStatus = sshSession->exitStatus();
@@ -594,6 +595,7 @@ void TerminalSession::onClosed()
         else
             sessionLog()("Process terminated after {} seconds.", diff.count());
     }
+#endif
     else
         sessionLog()("Process terminated after {} seconds.", diff.count());

diff --git a/src/contour/TerminalSessionManager.cpp b/src/contour/TerminalSessionManager.cpp
index a784529c..1dc2e534 100644
--- a/src/contour/TerminalSessionManager.cpp
+++ b/src/contour/TerminalSessionManager.cpp
@@ -4,7 +4,9 @@
 #include <contour/TerminalSessionManager.h>

 #include <vtpty/Process.h>
-#include <vtpty/SshSession.h>
+#if defined(VTPTY_LIBSSH2)
+    #include <vtpty/SshSession.h>
+#endif

 #include <QtQml/QQmlEngine>

@@ -26,8 +28,10 @@ std::unique_ptr<vtpty::Pty> TerminalSessionManager::createPty()
 {
     auto const& profile = _app.config().profile(_app.profileName());

+#if defined(VTPTY_LIBSSH2)
     if (!profile->ssh.hostname.empty())
         return make_unique<vtpty::SshSession>(profile->ssh);
+#endif

     return make_unique<vtpty::Process>(profile->shell, vtpty::createPty(profile->terminalSize, nullopt));
 }
```

My basic idea is: if there's no SSH client support found, we should not even compile it in.

Comment (Yaraslaut) [reply src/vtpty/CMakeLists.txt:47]: Yes, this also works, I will update pr

File: src/contour/display/OpenGLRenderer.h [hunk 2]
```diff
@@ -35,10 +35,13 @@ namespace contour::display
 {
 
 class OpenGLRenderer final:
+    public QObject,
     public vtrasterizer::RenderTarget,
     public vtrasterizer::atlas::AtlasBackend,
     public QOpenGLExtraFunctions
 {
+    Q_OBJECT
+
```
Comment (christianparpart) [inline src/contour/display/OpenGLRenderer.h:44]: Why was this readded?

**Pull request merged**